### PR TITLE
Jetpack E2E: update `Editor: Basic Post Flow` to also test Jetpack SEO & Social Previews.

### PR DESCRIPTION
**Pull request by worldomonation**
Related to https://github.com/Automattic/wp-calypso/issues/80730.

## Proposed Changes

This PR updates the flow mentioned in the title to also test the Jetpack specific features of SEO and Social Previews.

The SEO step is gated behind an environment variable check, but the social preview step is available for both Simple and AT environments.

## Testing Instructions

Ensure the following build configurations are passing:
  - [x] Calypso E2E (desktop)
  - [x] Jetpack E2E Simple (mobile)
  - [x] Jetpack E2E Simple (desktop)
  - [x] Jetpack E2E AT (desktop)

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts [hunk 1]
```diff
@@ -17,7 +12,7 @@ const selectors = {
 		`${ panel } button.is-active:has-text("${ tabName }")`,
 
 	// General section-related
-	section: ( name: ArticleSections ) =>
+	section: ( name: string ) =>
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:15]: I chose to remove `ArticleSections` type because it proved to be too restrictive against an indeterminate and quickly changing set of sidebar panel sections. 

For instance, there is a new section that was previously not present when this type was defined - `Newsletter` - and instead of type-gating, I will leave the onus of ensuring the correct heading to the caller.

File: packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts [hunk 2]
```diff
@@ -60,6 +55,41 @@ export class EditorSettingsSidebarComponent {
...
 	}
 
+	//#region Generic methods
+
+	/**
+	 * Clicks a button matching the accessible name.
+	 *
+	 * @param {string} name Accessible name of the button.
+	 */
+	async clickButton( name: string ): Promise< void > {
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:65]: The trend for now, at least for me, has been to implement more generic methods instead of tailored flows.

File: packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts [hunk 3]
```diff
@@ -94,20 +124,16 @@ export class EditorSettingsSidebarComponent {
...
 			return;
 		}
 
-		// Avoid the wpcalypso/staging banner.
-		await this.scrollToBottomOfSidebar();
-
 		const editorParent = await this.editor.parent();
 		const sectionLocator = editorParent.locator( selectors.section( name ) );
-		await sectionLocator.click();
+		await sectionLocator.click( { position: { x: 5, y: 5 } } );
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:138]: This effectively removes the need to call `scrollToBottomOfSidebar`, which hard-coded the expectation that a section called `Discussion` will be the last item in the settings panel.

That would have been brittle, plus for Jetpack settings panel, this won't work because the `Discussion` section does not exist there. So instead of forcing a scroll to the bottom, click on the bounding box at x:5 and y:5 to avoid the staging/wpcalypso badge entirely.

File: packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts [hunk 4]
```diff
@@ -268,15 +268,23 @@ export class EditorToolbarComponent {
...
+			const label = await this.translateFromPage( settingsButtonLabel );
+			const selector = selectors.settingsButton( label );
+
+			if ( await this.targetIsOpen( selector ) ) {
+				return;
+			}
+
+			locator = editorParent.locator( selector );
+		} else {
+			locator = editorParent.getByRole( 'button', { name: 'Jetpack', exact: true } );
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts]: This is the new portion of code. Everything else was untouched to avoid scope creep.

File: test/e2e/specs/editor/editor__post-basic-flow.ts [hunk 5]
```diff
@@ -77,6 +77,44 @@ describe( DataHelper.createSuiteTitle( 'Editor: Basic Post Flow' ), function ()
...
+	describe( 'Jetpack features', function () {
+		beforeAll( async function () {
+			await editorPage.openSettings();
+		} );
+
+		it( 'Open Jetpack post settings', async function () {
+			await editorPage.openSettings( 'Jetpack' );
+		} );
+
+		skipItIf( envVariables.TEST_ON_ATOMIC !== true )(
```
Comment (worldomonation) [inline test/e2e/specs/editor/editor__post-basic-flow.ts:92]: Jetpack settings > SEO is not available for non-AT sites.

Screenshot from a Simple test site.
image

File: test/e2e/specs/editor/editor__post-basic-flow.ts [hunk 6]
```diff
@@ -77,6 +77,44 @@ describe( DataHelper.createSuiteTitle( 'Editor: Basic Post Flow' ), function ()
...
+			await editorPage.clickSidebarButton( 'Open Social Previews' );
+		} );
+
+		it( 'Show social preview for Tumblr', async function () {
+			// Action implemented as "raw" calls for now (2023-09).
+			const editorParent = await editorPage.getEditorParent();
+			const dialog = editorParent.getByRole( 'dialog' );
+
+			await dialog.getByRole( 'tab', { name: 'Tumblr' } ).click();
+			await dialog.getByRole( 'tabpanel', { name: 'Tumblr' } ).waitFor();
```
Comment (worldomonation) [inline test/e2e/specs/editor/editor__post-basic-flow.ts:113]: I chose not to implement a POM for this because this specific type of dialog box is really only used in two places for Jetpack settings.

**Review by john-legg — CHANGES_REQUESTED** [review]
Hey @worldomonation 👋 Thanks for explaining everything and leaving detailed comments. I did a quick check of the builds and I noticed that the Jetpack Simple Deploy (mobile) failed on `editor__post-basic-flow` with a timeout, even after rerunning it. I have to EOD, so I didn't get a chance to investigate, but I just wanted to mention it!

Everything else looked good, even the Atomic build 👍

**Comment by worldomonation** [comment]
> Hey @worldomonation 👋 Thanks for explaining everything and leaving detailed comments. I did a quick check of the builds and I noticed that the Jetpack Simple Deploy (mobile) failed on `editor__post-basic-flow` with a timeout, even after rerunning it. I have to EOD, so I didn't get a chance to investigate, but I just wanted to mention it!
> 
> Everything else looked good, even the Atomic build 👍

Yeah, looks like something failed. Will address it today, thanks for taking a look!

**Comment by worldomonation** [comment]
@john-legg It is passing now. I will add explanations wherever required.

File: test/e2e/specs/editor/editor__post-basic-flow.ts [hunk 7]
```diff
@@ -64,8 +64,8 @@ describe( DataHelper.createSuiteTitle( 'Editor: Basic Post Flow' ), function ()
 	} );
 
 	describe( 'Categories and Tags', function () {
-		it( 'Open settings', async function () {
-			await editorPage.openSettings();
+		it( 'Open post settings', async function () {
```
Comment (worldomonation) [inline test/e2e/specs/editor/editor__post-basic-flow.ts:69]: Just a title change to make it clear we're dealing with post/page settings.

File: test/e2e/specs/editor/editor__post-basic-flow.ts [hunk 8]
```diff
@@ -75,15 +75,54 @@ describe( DataHelper.createSuiteTitle( 'Editor: Basic Post Flow' ), function ()
...
+
+			await dialog.getByRole( 'tab', { name: 'Tumblr' } ).click();
+			await dialog.getByRole( 'tabpanel', { name: 'Tumblr' } ).waitFor();
+		} );
+
+		it( 'Dismiss social preview', async function () {
+			await page.keyboard.press( 'Escape' );
+		} );
+
+		afterAll( async function () {
```
Comment (worldomonation) [inline test/e2e/specs/editor/editor__post-basic-flow.ts:128]: I've decided to remove the step `Close settings sidebar` and replace it with an `afterAll` hook. Closing the panel is not really a valid test on its own, so having it count as a step felt unnecessary.

**Review by john-legg — APPROVED** [review]

**Comment by john-legg** [comment]
Awesome! Everything looks good to me. Thanks for making the change 🚀

File: packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts (hunk 3 above)
Comment (dpasque) [reply packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:138]: This is such a great idea!! 🥳

File: packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts (hunk 4 above)
Comment (dpasque) [reply packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts]: **Nit (very-non-blocking):** This is totally fine, but I wonder if it might be cleaner to just implement this as a separate method -- having the two-option parameter and then just completely different paths in an if clause kind of makes me think this might be best as two separate methods. Especially because all sorts of stuff can be added by plugins to that toolbar! So it might grow a little smoother too.

File: test/e2e/specs/editor/editor__post-basic-flow.ts [hunk 9]
```diff
@@ -75,15 +75,54 @@ describe( DataHelper.createSuiteTitle( 'Editor: Basic Post Flow' ), function ()
...
 
-		// This step is required on mobile, but doesn't hurt anything on desktop, so avoiding conditional.
-		it( 'Close settings sidebar', async function () {
+		it( 'Show social preview for Tumblr', async function () {
+			// Action implemented as "raw" calls for now (2023-09).
+			const editorParent = await editorPage.getEditorParent();
+			const dialog = editorParent.getByRole( 'dialog' );
+
+			await dialog.getByRole( 'tab', { name: 'Tumblr' } ).click();
+			await dialog.getByRole( 'tabpanel', { name: 'Tumblr' } ).waitFor();
```
Comment (dpasque) [inline test/e2e/specs/editor/editor__post-basic-flow.ts:113]: **Suggestion (non-blocking)**: What do you think of going one small step further and actually validating the content of the preview? If we've added an SEO title, we can look for that text in there, otherwise we can look for the post title. I think that would be a nice final check, and I think we could do it pretty stably!

**Review by dpasque — APPROVED** [review]
Awesome -- I love the idea of rolling this into the Basic Post spec! 🚀 

I left a few suggestions and ideas for you, but this shouldn't need any re-review. Thanks for knocking this out! 😄

File: packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts (hunk 4 above)
Comment (worldomonation) [reply packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts]: Interestingly, I arrived at this solution because I felt having one method that could open any one of x different settings panels was easier than implementing a method per panel!

> having the two-option parameter and then just completely different paths in an if clause kind of makes me think this might be best as two separate methods. 

I did it this way because I didn't want to expand the scope of the changes, which is already moderately substantial relative to the spec itself. Lines 275 - 285 could probably be merged into one `getByRole` call and it _should_ just work. I'll make the change and verify.

File: test/e2e/specs/editor/editor__post-basic-flow.ts (hunk 9 above)
Comment (worldomonation) [reply test/e2e/specs/editor/editor__post-basic-flow.ts:113]: Sure, let me take a look at doing that - I initially chose not to check the rendered preview because it could potentially not generate, etc - and after being burned by the AI Assistant block, I was wary of testing implementation details 😆
